clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -252,6 +252,33 @@ public synchronized void update(Cluster newCluster, Set<String> unavailableTopic
         if (isClosed())
             throw new IllegalStateException("Update requested after metadata close");
 
+        String previousClusterId = cluster.clusterResource().clusterId();
+        String newClusterId = newCluster.clusterResource().clusterId();
+
+        if (previousClusterId != null && !newClusterId.equals(previousClusterId)) {
+            // kafka cluster id is unique.
+            // On client side, the cluster id in Metadata is only null during bootstrap, and client is
+            // expected to talk to the same cluster during its life cycle. Therefore if cluster id changes
+            // during metadata update, meaning this metadata update response is from a different cluster,
+            // client should reject this response and not update cached cluster to the wrong cluster
+
+            // According to SRE, removing brokers and adding to another cluster is common operation, thus
+            // it might be more suitable to just throw an exception for update and fail this update operation
+            // instead of bringing down the client completely, so that the metadata can be updated later from
+            // other brokers in the same cluster.
+
+            // remove the traitor broker from the cluster's node list
+            List<Node> originalNodes = new ArrayList<>(cluster.nodes());
+            for (Node node : newCluster.nodes()) {
+                originalNodes.remove(node);
```

Comment:
do we want to count the hits? we expect at least 1 ...

Reply:
do we also want to make sure we are left with >0 still-valid brokers ?

Reply:
sure